File: top-100-liked/94-binary-tree-inorder-traversal.ts
/**
 * 二叉树的中序遍历
 * https://leetcode.cn/problems/binary-tree-inorder-traversal/
 */

// Definition for a binary tree node.
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val
    this.left = left === undefined ? null : left
    this.right = right === undefined ? null : right
  }
}

export function inorderTraversal(root: TreeNode | null): number[] {
  const res = []

  const inorder = (root: TreeNode) => {
    if (!root) {
      return
    }
    inorder(root.left)
    res.push(root.val)
    inorder(root.right)
  }
  inorder(root)
  return res
}

File: leetcode/94-binary-tree-inorder-traversal.py
# 94. 二叉树的中序遍历
# https://leetcode.cn/problems/binary-tree-inorder-traversal
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.ans: List[int] = []

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return self.ans
        else:
            self.inorderTraversal(root.left)
            self.ans.append(root.val)
            self.inorderTraversal(root.right)
            return self.ans
